src/language/concurrency.md

```diff
@@ -144,80 +171,107 @@ void main() [!async!] {
+function body.
 {{site.alert.end}}
 
-As the following figure shows,
-the Dart code pauses while `readAsString()` executes non-Dart code,
-in either the Dart virtual machine (VM) or the operating system (OS).
+As the following figure shows, the Dart code pauses while `readAsString()`
+executes non-Dart code, in either the Dart runtime or the operating. 
 Once `readAsString()` returns a value, Dart code execution resumes.
 
-![Flowchart-like figure showing app code executing from start to exit, waiting for native I/O in between](/assets/img/language/concurrency/basics-await.png)
+![Flowchart-like figure showing app code executing from start to exit, waiting
+for native I/O in between](/assets/img/language/concurrency/basics-await.png)
 
-If you'd like to learn more about using `async`, `await`, and futures,
-visit the [asynchronous programming codelab][].
+### Streams
 
-[asynchronous programming codelab]: /codelabs/async-await
+Dart also supports asynchronous code in the form of streams. Streams
+provide values in the future and repeatedly over time. A promise to provide a
+series of `int` values over time has the type `Stream<int>`.
 
+In the following example, the stream created with `Stream.periodic`
+repeatedly emits a new `int` value every second.
 
-## How isolates work
+<?code-excerpt "lib/stream_syntax.dart"?>
+```dart
+Stream<int> stream = Stream.periodic(const Duration(seconds: 1), (i) => i * i);
+```
+
+#### await-for and yield
+
+Await-for is a type of for loop that executes each subsequent iteration of the
+loop as new values are provided. In other words, it’s used to “loop over”
+streams. In this example, a new value will be emitted from the function
+`sumStream` as new values are emitted from the stream that’s provided as an
+argument. The `yield` keyword is used rather than `return` in functions that 
+return streams of values.
+
+<?code-excerpt "lib/await_for_syntax.dart"?>
+```dart
+Stream<int> sumStream(Stream<int> stream) async* {
+  var sum = 0;
+  await for (final value in stream) {
+    yield sum += value;
+  }
+}
+```
+
+If you'd like to learn more about using `async`, `await`, `Stream`s and
+`Future`s,  visit the [asynchronous programming codelab][].
+
+[asynchronous programming codelab]: /codelabs/async-await
 
-Most modern devices have multi-core CPUs.
-To take advantage of multiple cores,
-developers sometimes use shared-memory threads running concurrently.
-However, shared-state concurrency is
-[error prone](https://en.wikipedia.org/wiki/Race_condition#In_software) and
-can lead to complicated code.
+## Isolates
 
-Instead of threads, all Dart code runs inside of isolates.
-Each isolate has its own memory heap,
-ensuring that none of the state in an isolate is accessible from
-any other isolate.
-No shared state between isolates means concurrency complexities like 
+Dart also supports concurrency via isolates. Most modern devices have multi-core
```

Comment:
Since this section probably will be linked to from around the site (and outside the site) as the defacto isolate definition, we can't assume the reader read anything before this so "also" doesn't really work. We can say:

 ```suggestion
Dart supports concurrency via isolates, in addition to [asynchronous APIs](#types-and-syntax-for-asynchronous-programming).
Most modern devices have multi-core
```

The link to the earlier section is, again, because this might be the first part of the page someone reads. 

(Idk if saying "asynchronous APIs" makes sense though so feel free to change that)